src/iterators.jl

```diff
@@ -1,145 +1,190 @@
-    celldofs::Vector{Int}
-
-    function CellIterator{dim,C,T}(dh::Union{DofHandler{dim,T,G},MixedDofHandler{dim,T,G},Nothing}, cellset::Union{AbstractVector{Int},Nothing}, flags::UpdateFlags) where {dim,C,T,G}
-        isconcretetype(C) || _check_same_celltype(dh.grid, cellset)
-        N = nnodes_per_cell(dh.grid, cellset === nothing ? 1 : first(cellset))
-        cell = ScalarWrapper(0)
-        nodes = zeros(Int, N)
-        coords = zeros(Vec{dim,T}, N)
-        n = ndofs_per_cell(dh, cellset === nothing ? 1 : first(cellset))
-        celldofs = zeros(Int, n)
-        return new{dim,C,T,typeof(dh)}(flags, dh.grid, cell, nodes, coords, cellset, dh, celldofs)
-    end
+    coords::Vector{X}
+    dh::DH
+    dofs::Vector{Int}
+end
 
-    function CellIterator{dim,C,T}(grid::Grid{dim,C,T}, cellset::Union{AbstractVector{Int},Nothing}, flags::UpdateFlags) where {dim,C,T}
-        isconcretetype(C) || _check_same_celltype(grid, cellset)
-        N = nnodes_per_cell(grid, cellset === nothing ? 1 : first(cellset))
-        cell = ScalarWrapper(0)
-        nodes = zeros(Int, N)
-        coords = zeros(Vec{dim,T}, N)
-        return new{dim,C,T,Nothing}(flags, grid, cell, nodes, coords, cellset)
-    end
+function CellCache(grid::Grid{dim,C,T}, flags::UpdateFlags=UpdateFlags()) where {dim,C,T}
+    N = nnodes_per_cell(grid)
+    nodes = zeros(Int, N)
+    coords = zeros(Vec{dim,T}, N)
+    return CellCache(flags, grid, ScalarWrapper(-1), nodes, coords, nothing, Int[])
+end
+
+function CellCache(dh::Union{DofHandler{dim,T},MixedDofHandler{dim,T}}, flags::UpdateFlags=UpdateFlags()) where {dim,T}
+    N = nnodes_per_cell(dh.grid)
+    nodes = zeros(Int, N)
+    coords = zeros(Vec{dim,T}, N)
+    n = ndofs_per_cell(dh)
+    celldofs = zeros(Int, n)
+    return CellCache(flags, dh.grid, ScalarWrapper(-1), nodes, coords, dh, celldofs)
 end
 
-CellIterator(grid::Grid{dim,C,T}, cellset::Union{AbstractVector{Int},Nothing}=nothing, flags::UpdateFlags=UpdateFlags()) where {dim,C,T} =
-    CellIterator{dim,C,T}(grid, cellset, flags)
-CellIterator(dh::DofHandler{dim,T}, cellset::Union{AbstractVector{Int},Nothing}=nothing, flags::UpdateFlags=UpdateFlags()) where {dim,T} =
-    CellIterator{dim,getcelltype(dh.grid),T}(dh, cellset, flags)
-CellIterator(dh::MixedDofHandler{dim,T}, cellset::Union{AbstractVector{Int},Nothing}=nothing, flags::UpdateFlags=UpdateFlags()) where {dim,T} =
-    CellIterator{dim,getcelltype(dh.grid),T}(dh, cellset, flags)
-
-# iterator interface
-function Base.iterate(ci::CellIterator, state = 1)
-    if state > (ci.cellset === nothing ? getncells(ci.grid) : length(ci.cellset))
-        return nothing
-    else
-        return (reinit!(ci, state), state+1)
+# TODO: Can always resize and combine the two reinit! methods maybe?
+function reinit!(cc::CellCache, i::Int)
+    cc.cellid[] = i
+    if cc.flags.nodes
+        cellnodes!(cc.nodes, cc.grid, i)
+    end
+    if cc.flags.coords
+        cellcoords!(cc.coords, cc.grid, i)
```

Comment:
Here you can change to `getcoordinates`, since `dh isa DofHandler` and has concrete cell type? And we will probably soon deprecate `cellcoords!`

Reply:
https://github.com/Ferrite-FEM/Ferrite.jl/blob/master/src/Grid/grid.jl#L668-L685 ?